### ISSUE TITLE
Missing predicate implementations for `json_path`

### ISSUE THREAD

**Issue by sfingram** [theredfish/grillon#87]
### Description

I'm getting a lot of great use out of grillon, and I think it is a terrific crate.  One thing I notice while testing with `json_path` assertions is that assertions like:
```rust
    Grillon::new(&mock_server.server.url("/"))?
        .get("users/1")
        .assert()
        .await
        .json_path("$.name", matches("Isa+c"));
```
Fail with:
```
not implemented: Invalid predicate for the json path DSL: should match
```
because only the `is`, `is_not`, `schema`, `contains`, `does_not_contain` have been implemented for `json_path`.  It would be fantastic if these other missing predicates could get implemented for `json_path` assertions.  Thanks!

### Workaround

In the meantime, I am using the following (janky) workaround with the `schema` predicate:
```rust
pub fn path_schema<T>(items: T) -> grillon::dsl::Expression<grillon::Value>
where
    T: serde::Serialize,
{
    grillon::dsl::schema(json!({
        "type": "array",
        "minItems": 1,
        "maxItems": 1,
        "items": items
    }))
}
```
Which can be used to implement `matches` like:
```rust
    Grillon::new(&mock_server.server.url("/"))?
        .get("users/1")
        .assert()
        .await
        .json_path("$.name", path_schema(json!({"type": "string", "pattern": "Isa+c"})));
```
Get the job done, but a true predicate would be the Right Way :tm:

**Comment by theredfish** [comment]
Thank you for the feedback @sfingram ! 

I'll take a look at it in the next few days. Glad to see that the `unimplemented` flow leads to an issue being opened :) . Not all predicates are compatible with built-in assertion functions, but this one is definitely a good candidate.

`json_path` needs more work - I've only implemented the basic needs. This kind of feedback helps with prioritization👍

**Comment by sfingram** [comment]
@theredfish Yes, that would cover it, thank you! 👍

**Comment by theredfish** [comment]
@sfingram can I ask you to give it a try by adding this commit to your Grillon dependency? `c33709b3e75332420573e42c6de3a1d258bacc2f` 

In your `Cargo.toml` replace `grillon` with this:

```toml
grillon = { git = "https://github.com/theredfish/grillon.git", rev = "c33709b3e75332420573e42c6de3a1d258bacc2f" }
```

Then you should have access to the `matches` and `does_not_match` predicates for the `json_path` assertion. I'm interested in the dev experience and possible edge cases. I wrote some tests but they are far from being exhaustive.

Some test cases could be:

- Checking numbers
- Dealing with null values
- Dealing with inexistant paths
- Arrays
- atomic values (string, number, ...)

I have some auth functionalities I would like to add for the `0.6.0`. Waiting for the release you can use this commit revision or any subsequent revision depending on your feedback. I'll communicate them when necessary.

But I shouldn't be too long for releasing 0.6.0

**Comment by theredfish** [comment]
To move forward this is now released under `v0.6.0` 

Here are two matching examples:

```rust
#[tokio::test]
async fn json_path_matches() -> Result<()> {
    let mock_server = HttpMockServer::new();
    let mock = mock_server.get_valid_user();

    Grillon::new(&mock_server.server.url("/"))?
        .get("users/1")
        .assert()
        .await
        .json_path("$.name", matches("Isa+c"))
        .json_path("$.name", matches(r"Isa[a-z]{2}"))
        .json_path("$.name", matches("Isaac"));

    mock.assert();

    Ok(())
}

#[tokio::test]
async fn json_path_does_not_match() -> Result<()> {
    let mock_server = HttpMockServer::new();
    let mock = mock_server.get_valid_user();

    Grillon::new(&mock_server.server.url("/"))?
        .get("users/1")
        .assert()
        .await
        .json_path("$.name", does_not_match("^Isa$"));

    mock.assert();

    Ok(())
}
```
